kore/src/Kore/Builtin/Map.hs

```diff
@@ -561,3 +581,109 @@ unifyEquals unifyEqualsChildren first second = do
             normalizedOrBottom
                 :: Ac.NormalizedOrBottom Domain.NormalizedMap variable
             normalizedOrBottom = Ac.toNormalized patt
+
+data InKeys term =
+    InKeys
+        { symbol :: !Symbol
+        , keyTerm, mapTerm :: !term
+        }
+
+matchInKeys :: TermLike variable -> Maybe (InKeys (TermLike variable))
+matchInKeys (App_ symbol [keyTerm, mapTerm]) = do
+    hook2 <- (getHook . symbolHook) symbol
+    Monad.guard (hook2 == Map.in_keysKey)
+    return InKeys { symbol, keyTerm, mapTerm }
+matchInKeys _ = Nothing
+
+unifyNotInKeys
+    :: forall variable unifier
+    .  InternalVariable variable
+    => MonadUnify unifier
+    => TermSimplifier variable unifier
+    -> NotSimplifier unifier
+    -> CeilSimplifier unifier (TermLike variable) (OrCondition variable)
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+unifyNotInKeys
+    unifyChildren
+    (NotSimplifier notSimplifier)
+    (CeilSimplifier ceilSimplifier)
+    a
+    b
+  =
+    worker a b <|> worker b a
+  where
+    normalizedOrBottom
+       :: TermLike variable
+       -> Ac.NormalizedOrBottom Domain.NormalizedMap variable
+    normalizedOrBottom = Ac.toNormalized
+
+    defineTerm :: TermLike variable -> MaybeT unifier (Condition variable)
+    defineTerm termLike =
+        ceilSimplifier
+            TermLike.Ceil
+                { ceilResultSort = Sort.predicateSort
+                , ceilOperandSort = TermLike.termLikeSort termLike
+                , ceilChild = termLike
+                }
+        >>= lift . Unify.scatter
+
+    eraseTerm =
+        Pattern.fromCondition . Pattern.withoutTerm
+
+    unifyAndNegate t1 t2 = do
+        unificationSolutions <-
```

Comment:
```suggestion
        -- Erasing the unified term is valid here because
        -- the terms are all wrapped in \ceil below.
        unificationSolutions <-
```